interfaces/builtin/custom_device.go

```diff
@@ -0,0 +1,367 @@
+	}
+
+	if err := iface.validateFilePath(path); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (iface *customDeviceInterface) validatePaths(slot *snap.SlotInfo, attrName string) error {
+	var paths []string
+	if err := slot.Attr(attrName, &paths); err != nil {
+		if !errors.Is(err, snap.AttributeNotFoundError{}) {
+			return err
+		}
+	}
+
+	for _, path := range paths {
+		if err := iface.validateFilePath(path); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (iface *customDeviceInterface) validateUDevValue(value interface{}) error {
+	stringValue, ok := value.(string)
+	if !ok {
+		return fmt.Errorf(`value "%v" is not a string`, value)
+	}
+
+	if !customDeviceUDevValueRegexp.MatchString(stringValue) {
+		return fmt.Errorf(`value "%v" contains invalid characters`, stringValue)
+	}
+
+	return nil
+}
+
+func (iface *customDeviceInterface) validateUDevValueMap(value interface{}) error {
+	valueMap, ok := value.(map[string]interface{})
+	if !ok {
+		return fmt.Errorf(`value "%v" is not a map`, value)
+	}
+
+	for key, val := range valueMap {
+		if !customDeviceUDevValueRegexp.MatchString(key) {
+			return fmt.Errorf(`key "%v" contains invalid characters`, key)
+		}
+		if err := iface.validateUDevValue(val); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (iface *customDeviceInterface) validateUDevTaggingRule(rule map[string]interface{}, devices []string) error {
+	hasKernelTag := false
+	for key, value := range rule {
+		var err error
+		switch key {
+		case "subsystem":
+			err = iface.validateUDevValue(value)
+		case "kernel":
+			hasKernelTag = true
+			err = iface.validateUDevValue(value)
+			if err == nil {
+				deviceName := value.(string)
+				// furthermore, the kernel name must match the name of one of
+				// the given devices
+				if !strutil.ListContains(devices, "/dev/"+deviceName) {
+					err = fmt.Errorf(`%q does not match a specified device`, deviceName)
+				}
+			}
+		case "attributes", "environment":
+			err = iface.validateUDevValueMap(value)
+		default:
+			err = errors.New(`unknown key`)
+		}
+
+		if err != nil {
+			return fmt.Errorf(`custom-device "udev-tagging" invalid %q tag: %v`, key, err)
+		}
+	}
+
+	if !hasKernelTag {
+		return errors.New(`custom-device udev tagging rule missing mandatory "kernel" key`)
+	}
+
+	return nil
+}
+
+func (iface *customDeviceInterface) Name() string {
+	return "custom-device"
+}
+
+func (iface *customDeviceInterface) StaticInfo() interfaces.StaticInfo {
+	return interfaces.StaticInfo{
+		Summary:              customDeviceSummary,
+		BaseDeclarationPlugs: customDeviceBaseDeclarationPlugs,
+		BaseDeclarationSlots: customDeviceBaseDeclarationSlots,
+		AffectsPlugOnRefresh: true,
+	}
+}
+
+func (iface *customDeviceInterface) BeforePrepareSlot(slot *snap.SlotInfo) error {
+	if slot.Attrs == nil {
+		slot.Attrs = make(map[string]interface{})
+	}
+	// custom-device defaults to "slot" name if unspecified
+	slot.Attrs["custom-device"] = slot.Name
```

Comment:
if I compare with content code, this is missing an if. Maybe it works anyway but it's not well defined to have both static and dynamic attributes of the same value?

do we have tests that set custom-device?

Reply:
I just decided to keep things simple in the slot side, and support only
```
slots:
  my-slot-name:
    interface: custom-device
```
instead of
```
slots:
  some-other-name:
    custom-device: my-slot-name
    interface: custom-device
```

I assume that we can have more leverage over gadgets authors and can be more restrictive with them about style issues; whereas on the plug side I left the `if`, so both styles are possible for snap authors. But let me know, if you prefer to allow both styles on the slots too.

Reply:
we don't do this anywhere else afaik, I think it's mostly confusing so -1. I think the slot should behave like the plug and allow for custom-device to be optionally explicit

Reply:
It looks like I missed this comment, before. I've now updated the logic to behave just like the plug.